openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMSWIG.java

```diff
@@ -144,7 +144,42 @@ private void copyDataToSWIGInstance(final Instance instance) {
         }
     }
 
+    /**
+     * Returns the features contributions for the current instance.
+     *
+     * @param instance The instance from pulse.
+     * @return double[] array with features contributions.
+     * @since 1.2.2
+     */
+    public double[] getFeaturesContributions(final Instance instance) {
 
+        // we need to lock the resources to avoid having multiple threads sharing the same swig resources.
+        synchronized (this.swigResources) {
```

Comment:
I may be saying nonsense, but could this cause a performance bottleneck when computing feature contribution on the critical path?
I am aware it is happening for classification as well

Reply:
that's true, we will have to wait for the contributions. But it is not the lock that causes the problem of bottleneck, but the computation itself.

Reply:
do we need to do the same block operation for scoring and feature contributions?

Reply:
AFAIK, Lightgbm provides four predicts, two of them are `[C_API_PREDICT_NORMAL](https://lightgbm.readthedocs.io/en/latest/C-API.html#c.C_API_PREDICT_NORMAL)` and `[C_API_PREDICT_CONTRIB](https://lightgbm.readthedocs.io/en/latest/C-API.html#c.C_API_PREDICT_CONTRIB)`.
I didn't see the possibility of predicting the both values by an single call. I believe either by @SamuelNLP at the time that he had worked on this.
Nonetheless, I'm going to create a task to see how we can remove the lock so we can reduce bottleneck, since it is predicting one instance at the time.